File: frontend/types/api-response.ts
export class ApiResponse<
  TData = Record<string, any>,
  TError = Record<string, any>,
> {
  statusCode: number;
  data: TData;
  error: TError;
  message: string;
  isSuccess: boolean;

  constructor(
    statusCode: number = 200,
    data: TData | TError = {} as TData,
    message: string = "Success",
  ) {
    this.isSuccess = statusCode < 400;
    this.statusCode = statusCode;
    this.data = this.isSuccess ? (data as TData) : ({} as TData);
    this.error = this.isSuccess ? ({} as TError) : (data as TError);
    this.message = message;
  }
}

File: backend/src/utils/api_response.py
from typing import Generic, TypeVar

from pydantic import BaseModel

TData = TypeVar("TData")
TError = TypeVar("TError")


class ApiResponse(BaseModel, Generic[TData, TError]):
    statusCode: int
    data: TData
    error: TError
    message: str
    isSuccess: bool

    @classmethod
    def success(
        cls, data: TData = {}, message: str = "Success", status_code: int = 200
    ):
        return cls(
            statusCode=status_code,
            data=data,
            error={},
            message=message,
            isSuccess=True,
        )

    @classmethod
    def failure(
        cls, error: TError = {}, message: str = "Error", status_code: int = 500
    ):
        return cls(
            statusCode=status_code,
            data={},
            error=error,
            message=message,
            isSuccess=False,
        )
